hooks/useTransactionSender.tsx

```diff
@@ -1,25 +1,320 @@
+          stroke={2}
+          style={{ width: '1.25rem', height: '1.25rem' }}
+          color="var(--mantine-color-yellow-filled)"
+        />
+      );
+    }
+
+    return <Loader size="xs" color="var(--mantine-color-dark-2)" />;
+  };
+
+  /**
+   * asynchronously confirm transaction result for each signature and update state when resolved
+   */
+  const confirmTransaction = async (transactionInfos: Array<TransactionInfo<T>>, id: string) => {
+    transactionInfos.forEach(async (transactionInfo) => {
+      const controller = new AbortController();
+      const signal = controller.signal;
+
+      const { blockhash, lastValidBlockHeight } = await connection.getLatestBlockhash();
+      const confirmTransactionPromise = connection.confirmTransaction({
+        signature: transactionInfo.signature,
+        blockhash,
+        lastValidBlockHeight,
+      });
+
+      const timeoutPromise = new Promise((_, reject) => {
+        /**
+         * note: we use a constant timeout of 30s (seems standard across many solana projects) for simplicity
+         * for now. in the future, we might opt to dynamically set this timeout based on various cluster stats,
+         * like ping time
+         */
+        const timeoutId = setTimeout(() => {
+          controller.abort();
+        }, 30_000);
+
+        signal.addEventListener('abort', () => {
+          clearTimeout(timeoutId);
+          reject(new Error('Transaction confirmation aborted'));
+        });
+      });
+
+      Promise.race([confirmTransactionPromise, timeoutPromise])
+        .then((result) => {
+          if ((result as any).value) {
+            const transactionHasError =
+              (result as RpcResponseAndContext<SignatureResult>).value.err !== null;
+
+            updateSingleTransactionInfo(transactionInfo.signature, id, () => ({
+              result: {
+                err: transactionHasError ? {} : null,
+              },
+            }));
+
+            setSuccessfulSignatureCount((state) => {
+              const current = state[id] ?? 0;
+              return {
+                ...state,
+                [id]: transactionHasError ? current : current + 1,
+              };
+            });
+          }
+        })
+        .catch((err) => {
+          console.error(
+            `an error occurred resolving transaction for signature [${transactionInfo.signature}]`,
+            err,
+          );
+
+          updateSingleTransactionInfo(transactionInfo.signature, id, () => ({
+            status: TransactionStatus.TIMED_OUT,
+          }));
+        });
+    });
+  };
 
   const send = useCallback(
     /**
      * Sends transactions.
      * @param txs A sequence of sets of transactions. Sets are executed simultaneously.
      * @returns A sequence of set of tx signatures.
      */
-    async <T extends Transaction | VersionedTransaction>(txs: SingleOrArray<T>[]) => {
+    async (txs: SingleOrArray<T | TransactionWithMetadata<T>>[]) => {
```

Comment:
Is there an opportunity to break out part of this function into a separate function so it's a little more readable/maintainable? No worries if not just wanted  to see.

Reply:
Fair point, it's definitely quite long — I think there are some parts I can pull out. Will give that a shot and see what you think.

Reply:
I'm also not totally sure what the point of maintaining the sequences and sets throughout the function is since there's no guarantees (AFAIK) around transaction ordering after you actually send them to the RPC — unless of course, we were to use Jito bundles.

I only say this to point out it's requiring more code, but I don't want to change something I don't understand, so I'll leave it for now.

Reply:
@jshiohaha hey that totally works for me. Let's leave it as is.